### ISSUE TITLE
[STORM-2382] log4j and slf4j conflicting libraries issue

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#6164]
<p>my project storm 1.0.1 job's dependencies (log4j & slf4j) conflict with apache&hdp storm 1.0.1 default libraries (STORM_HOME/lib) and is preventing submitting new storm job.</p>

<ul>
	<li>1. shadow my storm job jar *<br/>
I get the following error:</li>
</ul>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">SLF4J: <span class="code-object">Class</span> path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/hdp/2.5.0.0-1245/storm/lib/log4j-slf4j-impl-2.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/tmp/356865dafc1a11e69341ecb1d7ac1510.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http:<span class="code-comment">//www.slf4j.org/codes.html#multiple_bindings <span class="code-keyword">for</span> an explanation.
</span>SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]
Exception in thread <span class="code-quote">"main"</span> java.lang.IllegalAccessError: tried to access method org.apache.logging.log4j.core.lookup.MapLookup.newMap(I)Ljava/util/HashMap; from <span class="code-keyword">class </span>org.apache.logging.log4j.core.lookup.MainMapLookup
	at org.apache.logging.log4j.core.lookup.MainMapLookup.<clinit>(MainMapLookup.java:37)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.logging.log4j.core.util.ReflectionUtil.instantiate(ReflectionUtil.java:185)
	at org.apache.logging.log4j.core.lookup.Interpolator.<init>(Interpolator.java:65)
	at org.apache.logging.log4j.core.config.AbstractConfiguration.doConfigure(AbstractConfiguration.java:346)
	at org.apache.logging.log4j.core.config.AbstractConfiguration.start(AbstractConfiguration.java:161)
	at org.apache.logging.log4j.core.LoggerContext.setConfiguration(LoggerContext.java:359)
	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:420)
	at org.apache.logging.log4j.core.LoggerContext.start(LoggerContext.java:138)
	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:147)
	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:41)
	at org.apache.logging.log4j.LogManager.getContext(LogManager.java:175)
	at org.apache.logging.log4j.spi.AbstractLoggerAdapter.getContext(AbstractLoggerAdapter.java:102)
	at org.apache.logging.slf4j.Log4jLoggerFactory.getContext(Log4jLoggerFactory.java:43)
	at org.apache.logging.log4j.spi.AbstractLoggerAdapter.getLogger(AbstractLoggerAdapter.java:42)
	at org.apache.logging.slf4j.Log4jLoggerFactory.getLogger(Log4jLoggerFactory.java:29)
	at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:277)
	at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:288)
	at org.apache.storm.utils.LocalState.<clinit>(LocalState.java:45)
	at java.lang.<span class="code-object">Class</span>.forName0(Native Method)
	at java.lang.<span class="code-object">Class</span>.forName(<span class="code-object">Class</span>.java:348)
	at clojure.lang.RT.classForName(RT.java:2154)
	at clojure.lang.RT.classForName(RT.java:2163)
	at org.apache.storm.config__init.__init7(Unknown Source)
	at org.apache.storm.config__init.<clinit>(Unknown Source)
	at java.lang.<span class="code-object">Class</span>.forName0(Native Method)
	at java.lang.<span class="code-object">Class</span>.forName(<span class="code-object">Class</span>.java:348)
	at clojure.lang.RT.classForName(RT.java:2154)
	at clojure.lang.RT.classForName(RT.java:2163)
	at clojure.lang.RT.loadClassForName(RT.java:2182)
	at clojure.lang.RT.load(RT.java:436)
	at clojure.lang.RT.load(RT.java:412)
	at clojure.core$load$fn__5448.invoke(core.clj:5866)
	at clojure.core$load.doInvoke(core.clj:5865)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invoke(core.clj:5671)
	at clojure.core$load_lib$fn__5397.invoke(core.clj:5711)
	at clojure.core$load_lib.doInvoke(core.clj:5710)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invoke(core.clj:632)
	at clojure.core$load_libs.doInvoke(core.clj:5753)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invoke(core.clj:634)
	at clojure.core$use.doInvoke(core.clj:5843)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at org.apache.storm.command.config_value$loading__5340__auto____12764.invoke(config_value.clj:16)
	at org.apache.storm.command.config_value__init.load(Unknown Source)
	at org.apache.storm.command.config_value__init.<clinit>(Unknown Source)
	at java.lang.<span class="code-object">Class</span>.forName0(Native Method)
	at java.lang.<span class="code-object">Class</span>.forName(<span class="code-object">Class</span>.java:348)
	at clojure.lang.RT.classForName(RT.java:2154)
	at clojure.lang.RT.classForName(RT.java:2163)
	at clojure.lang.RT.loadClassForName(RT.java:2182)
	at clojure.lang.RT.load(RT.java:436)
	at clojure.lang.RT.load(RT.java:412)
	at clojure.core$load$fn__5448.invoke(core.clj:5866)
	at clojure.core$load.doInvoke(core.clj:5865)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.lang.Var.invoke(Var.java:379)
	at org.apache.storm.command.config_value.<clinit>(Unknown Source)
</pre>
</div></div>

<ul>
	<li>2. upgrade STORM_HOME/lib *<br/>
I get this warning.  looks like a racing condition <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/help_16.png" height="16" width="16" align="absmiddle" alt="" border="0"/></li>
</ul>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">log4j:WARN No appenders could be found <span class="code-keyword">for</span> logger (org.apache.storm.utils.Utils).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http:<span class="code-comment">//logging.apache.org/log4j/1.2/faq.html#noconfig <span class="code-keyword">for</span> more info.
</span>log4j:WARN No appenders could be found <span class="code-keyword">for</span> logger (org.apache.storm.utils.Utils).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http:<span class="code-comment">//logging.apache.org/log4j/1.2/faq.html#noconfig <span class="code-keyword">for</span> more info.</span>
</pre>
</div></div>

---
<details><summary><i>Originally reported by <a title="Sree Vaddi" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sreevaddi">sreevaddi</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-2382" target="_blank">log4j and slf4j conflicting libraries issue</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2024-07-02T10:47:15+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sreevaddi">sreevaddi</a>:</i>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">$ ls -l /usr/hdp/2.5.0.0-1245/storm/lib
total 27632
-rwxr-xr-x 1 root root    53231 Mar 28 07:53 asm-5.0.3.jar
-rwxr-xr-x 1 root root  3837631 Mar 28 07:53 clojure-1.7.0.jar
-rwxr-xr-x 1 root root    79474 Mar 28 07:53 disruptor-3.3.2.jar
-rwxr-xr-x 1 root root   285211 Mar 28 07:53 kryo-3.0.3.jar
-rwxr-xr-x 1 root root    29257 Mar 28 07:53 slf4j-api-1.7.7.jar
-rwxr-xr-x 1 root root   133531 Mar 28 07:53 log4j-api-2.1.jar
-rwxr-xr-x 1 root root   824749 Mar 28 07:53 log4j-core-2.1.jar
-rwxr-xr-x 1 root root    20647 Mar 28 07:53 log4j-over-slf4j-1.6.6.jar
-rwxr-xr-x 1 root root    22542 Mar 28 07:53 log4j-slf4j-impl-2.1.jar
-rwxr-xr-x 1 root root     5711 Mar 28 07:53 minlog-1.3.0.jar
-rwxr-xr-x 1 root root    41755 Mar 28 07:53 objenesis-2.1.jar
-rwxr-xr-x 1 root root    20738 Mar 28 07:53 reflectasm-1.10.1.jar
-rwxr-xr-x 1 root root     5647 Mar 28 07:53 ring-cors-0.1.5.jar
-rwxr-xr-x 1 root root   105112 Mar 28 07:53 servlet-api-2.5.jar
-rwxr-xr-x 1 root root 20303157 Mar 28 07:53 storm-core-1.0.1.2.5.0.0-1245.jar
-rwxr-xr-x 1 root root  2492576 Mar 28 07:53 storm-rename-hack-1.0.1.2.5.0.0-1245.jar
lrwxrwxrwx 1 root root       45 Mar 28 07:55 zookeeper.jar -> /usr/hdp/2.5.0.0-1245/zookeeper/zookeeper.jar
</pre>
</div></div>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sreevaddi">sreevaddi</a>:</i>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">slf4j-api-1.7.7.jar
log4j-api-2.1.jar
log4j-core-2.1.jar
log4j-over-slf4j-1.6.6.jar
log4j-slf4j-impl-2.1.jar
</pre>
</div></div>
<p>the main issue is, <tt>log4j-over-slf4j-1.6.6.jar</tt> and <tt>log4j-slf4j-impl-2.1.jar</tt>, living together.<br/>
refer to: <a href="https://logging.apache.org/log4j/2.x/log4j-slf4j-impl/index.html" class="external-link" target="_blank" rel="nofollow noopener">https://logging.apache.org/log4j/2.x/log4j-slf4j-impl/index.html</a></p>

<p>if upgraded also needs, <tt>log4j-1.2.17.jar</tt>, due to direct reference to it in the storm code base.</p>

<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz" class="user-hover" rel="ptgoetz">P. Taylor Goetz</a><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha" class="user-hover" rel="sriharsha">Harsha</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=antoine.tran">antoine.tran</a>:</i>
<p>What we did in our project was to exclude sf4j bridge to log4J and vice-versa everywhere. This way, Storm own logging framework continue to work, and we put a separated log4j contiguration in Storm classpath.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<p>Hi <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sreevaddi" class="user-hover" rel="sreevaddi">Sree Vaddi</a>. What you are seeing looks to be because there are multiple SLF4J bindings on the classpath. Here is a brief description of the jars involved:</p>

<ul>
	<li>slf4j-api - Pretty self-explanatory, this is the SLF4J API classes. There's no implementation included in this.</li>
	<li>log4j-sfl4j-impl-2.1 - This is the Log4j 2 binding to SLF4J. This causes logs via the SLF4J API to go to a Log4j 2 logger. These loggers are configured via the log4j2/worker.xml configuration file.</li>
	<li>log4j-over-slf4j-1.6.6 - This is the Log4j 1 bridge to SLF4j. This causes logs on the Log4j 1 API to go to SLF4J. The result of this file along with log4j-slf4j-impl-2.1 is that both SLF4j API and Log4j 1 API logs will go to the Log4j 2 API.</li>
</ul>


<p>Then there are the other two files, log4j-api-2.1 and log4j-core-2.1, which contain the Log4j 2 API and implementation. </p>

<p>Given this line</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">SLF4J: Found binding in [jar:file:/tmp/356865dafc1a11e69341ecb1d7ac1510.jar!/org/slf4j/impl/StaticLoggerBinder.class]
</pre>
</div></div>
<p>it looks like your topology jar contains a different SLF4J binding from log4j-sfl4j-impl-2.1. You should ensure that your topology jar doesn't depend on the following (not a complete list):</p>

<ul>
	<li>slf4j-log4j12 - This is the Log4j 1 SLF4J binding.</li>
	<li>slf4j-simple - This is a simple stdout binding for SLF4J</li>
</ul>


<p>You may also want to check that your jar doesn't depend on log4j.jar, the Log4j1 implementation. log4j-over-slf4j-1.6.6 will provide classes that conflict with this.</p>

<p>Finally you should set the jars provided by Storm to "provided" scope in Maven. You don't want your jar to contain classes that are also on the Storm classpath. Someone else ran into this here, maybe it'll be helpful: <a href="https://stackoverflow.com/questions/36097878/exception-in-thread-main-java-lang-illegalaccesserror-in-storm-topology" class="external-link" target="_blank" rel="nofollow noopener">https://stackoverflow.com/questions/36097878/exception-in-thread-main-java-lang-illegalaccesserror-in-storm-topology</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Lower the priority since it is specific to the case and answer is provided. I'm thinking about closing this if there's no further input here.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sreevaddi">sreevaddi</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a>:</p>

<p>Thank you. But I do not have the choice of changing the jar in my topology due to other dependencies.</p>

<p>I upgraded and made our own internal storm jar (I know) and proceeded further.</p>

<p> </p>

<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a>:</p>

<p>Please consider removing this tight coupling of logging, either now or in the future.</p>

<p> </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rzo1">rzo1</a>:</i>
<p>Storm 1.x isn't maintained anymore. If the issue still applies for 2.6.x, feel free to re-open.</p>
